datajunction-server/datajunction_server/internal/nodes.py

```diff
@@ -1317,3 +1325,213 @@ def get_cube_revision_metadata(session: Session, name: str):
+        f"ON {link_input.join_on}",
+    )
+    exc = DJException()
+    ctx = ast.CompileContext(session=session, exception=exc)
+    join_query.compile(ctx)
+    join_relation = join_query.select.from_.relations[0].extensions[0]  # type: ignore
+
+    # Verify that the query references both the node and the dimension being joined
+    expected_references = {node_name, link_input.dimension_node}
+    references = {
+        table.name.namespace.identifier()  # type: ignore
+        for table in join_relation.criteria.on.find_all(ast.Column)  # type: ignore
+    }
+    if expected_references.difference(references):
+        raise DJInvalidInputException(
+            f"The join SQL provided does not reference both the origin node {node_name} and the "
+            f"dimension node {link_input.dimension_node} that it's being joined to.",
+        )
+
+    # Verify that the columns in the ON clause exist on both nodes
+    if ctx.exception.errors:
+        raise DJException(
+            message=f"Join query {link_input.join_on} is not valid",
+            errors=ctx.exception.errors,
+        )
+
+    # Find an existing dimension link if there is already one defined for this node
+    existing_link = [
+        link
+        for link in node.current.dimension_links
+        if link.dimension_id == dimension_node.id and link.role == link_input.role
+    ]
+    is_update = False
+
+    if existing_link:
+        # Update the existing dimension link
+        is_update = True
+        dimension_link = existing_link[0]
+        dimension_link.join_sql = link_input.join_on
+        dimension_link.join_type = DimensionLink.parse_join_type(
+            join_relation.join_type,
+        )
+        dimension_link.join_cardinality = link_input.join_cardinality
+    else:
+        # If there is no existing link, create new dimension link object
+        dimension_link = DimensionLink(
+            node_revision_id=node.current.id,
+            dimension_id=dimension_node.id,
+            join_sql=link_input.join_on,
+            join_type=DimensionLink.parse_join_type(join_relation.join_type),
+            join_cardinality=link_input.join_cardinality,
+            role=link_input.role,
+        )
+
+    # Add/update the dimension link in the database
+    session.add(dimension_link)
+    session.add(
+        History(
+            entity_type=EntityType.LINK,
+            entity_name=node.name,
+            node=node.name,
+            activity_type=ActivityType.CREATE if not is_update else ActivityType.UPDATE,
+            details={
+                "dimension": dimension_node.name,
+                "join_sql": link_input.join_on,
+                "join_cardinality": link_input.join_cardinality,
+                "role": link_input.role,
+            },
+            user=current_user.username if current_user else None,
+        ),
+    )
+    session.commit()
+    session.refresh(node)
+    return is_update
```

Comment:
I am not sure this Boolean value is a useful return. I think returning nothing would be cleaner (like in `remove_dimension_link`), or returning an Enum (NEW, UPDATED, etc), but this true/false may just catch us by surprise somewhere.

Reply:
Makes sense! I think it's easier to just return an `ActivityType` enum, since we already have this enum for the history APIs. Let me make this change